### ISSUE TITLE
SSH mode comma-separated list of hosts does not parse hostnames

### ISSUE THREAD

**Issue by drkrynstrng** [glentner/hyper-shell#12]
The docs mention a comma-separated list of hosts can be used when using ssh mode across multiple nodes. However, it seems the hostnames are not parsed correctly. Using HyperShell 2.1.0 installed with Python 3.9.12, I get the following error:

```
$ hyper-shell cluster helloworld.txt -N 4 --ssh a03-07,a03-08
 WARNING [hypershell.server] No database configured - automatically disabled
   DEBUG [hypershell.server] Started
   DEBUG [hypershell.submit] Started (helloworld.txt)
   DEBUG [hypershell.server] Started (heartbeat)
   DEBUG [hypershell.server] Started (receiver)
   DEBUG [hypershell.submit] Started (loader)
   DEBUG [hypershell.submit] Started (committer: no database)
   DEBUG [hypershell.cluster] Launching client: ssh   a03-07,a03-08  /home/user/.local/bin/hyper-shell client -H a03-07.hpc.usc.edu -p 50001 -N 4 -b 1 -w 5 -t '"{}"' -k e16499c2d1d01707e65ee97c97a683204f5388b81071423cd27d51bd14b239eec589907ae70d1f2fc68bbb0ee306b6f798d3d009bb96f054b3e2bcba2a18ab7c -d 0
ssh: Could not resolve hostname a03-07,a03-08: Name or service not known
```

In this case I can successfully use the expandable pattern instead (a03-[07-08]).

For a Slurm job, ideally we could set ssh hosts using the environment variable SLURM_JOB_NODELIST, the value of which could be a combination of comma-separation and expandable pattern.

**Comment by glentner** [comment]
Hello @drkrynstrng, thank you for posting the issue.

Indeed, `a03-07` would not be parsed as a range; but `a[03-07]` would be expanded correctly. There is room for improvement here of course in dealing with ambiguities with the `-`, `,`, and `[]` notation. The comma-separated list behavior and range behavior are bracketed within the `[]` characters as with some other command-line tools.

Regarding Slurm clusters, and `SLURM_JOB_NODELIST`. You could indeed probably parse one or more of the environment variables set by the scheduler within your job context to construct a node list. However, if you have the benefit of a Slurm job context, why not use `--launcher=srun` instead an SSH based scheme.

In the near future the project can/will be improved with more robust guard rails and improved testing. I'm happy to consider any feedback from a usability perspective.

**Comment by glentner** [comment]
I think we should allow for comma-separated hostnames outside the confines of the concept of a range. E.g., `a[01-03,04,08]` expands correctly to a01, a02, a03, 04, and a08. But `host1,host2` does not expand to host1 and host2.

In the meantime, you might consider adding something like this to your configuration file to make your life easier.
Then use `--ssh --ssh-group=mygroup`.

```toml
[ssh.nodelist]
mygroup = ["host1", "host2", "host3"]
```

**Comment by drkrynstrng** [comment]
I see, thanks. To clarify, in my example there are two nodes: node 1 is named a03-07 and node 2 is named a03-08 (it denotes rack-node in rack).

I'm also testing srun and mpirun but I keep encountering an issue where the tasks complete but the HyperShell server does not shut down (it seems that some of the clients do not disconnect) and so the Slurm job does not end until timeout. That's a different issue, but I will debug further and create a separate GitHub issue if needed.

**Comment by glentner** [comment]
Let me know if you find anything with regard to the mpirun/srun workflow not shutting down. I'll also include something more flexible and robust for hostname declarations in the next release - should be soon as I have stuff on develop that I want to get out soon.

**Comment by drkrynstrng** [comment]
I was getting confused by the `--num-tasks` option and the optimal value to use. If I do not use this option, then using `srun` or `mpirun` as the launcher works fine. The resources are fully utilized and the job completes shortly after all tasks complete.

**Comment by glentner** [comment]
May be a bit longer before a production release makes it into the wild. But the `develop` branch now has a commit that includes a fix for this behavior.

Now we separate on commas _outside_ brackets as distinct groupings, and within those groupings you have the range-based expansions. E.g.,

```sh
$ hyper-shell cluster ... --ssh 'a000.local,b00[1-4,8].local'
```

Will now expand correctly into
```
a000.local
b001.local
b002.local
b003.local
b004.local
b008.local
```

**Comment by drkrynstrng** [comment]
Thanks, that works for me.

**Comment by glentner** [comment]
Just published v2.2.0 earlier today ([release notes](https://hyper-shell.readthedocs.io/en/latest/blog/20230329_2_2_0_release.html)). You should be able to upgrade normally without needing to checkout the develop branch or anything crazy.
